drivers/bbram/bbram_stm32.c

```diff
@@ -0,0 +1,107 @@
+/*
+ * Copyright (c) 2022 Google Inc
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#define DT_DRV_COMPAT st_stm32_bbram
+
+#include <errno.h>
+
+#include <zephyr/drivers/bbram.h>
+#include <zephyr/logging/log.h>
+#include <zephyr/sys/util.h>
+LOG_MODULE_REGISTER(bbram, CONFIG_BBRAM_LOG_LEVEL);
+
+#define STM32_BKP_REG_BYTES		 4
+#define STM32_BKP_REG_OFFSET		 0x50
+#define STM32_BKP_REG_INDEX(offset)	 ((offset) >> 2)
+#define STM32_BKP_REG_BYTE_INDEX(offset) ((offset)&0x3UL)
+#define STM32_BKP_REG(i)		 (((volatile uint32_t *)config->base_addr)[(i)])
+
+/** Device config */
+struct bbram_stm32_config {
+	const struct device *parent;
+	/* BBRAM base address */
+	uintptr_t base_addr;
+	/* BBRAM size in bytes. */
+	int size;
+};
+
+static int bbram_stm32_read(const struct device *dev, size_t offset, size_t size, uint8_t *data)
+{
+	const struct bbram_stm32_config *config = dev->config;
+	uint32_t reg, begin, to_copy;
+
+	if (size < 1 || offset + size > config->size) {
+		return -EFAULT;
+	}
+
+	for (size_t read = 0; read < size; read += to_copy) {
+		reg = STM32_BKP_REG(STM32_BKP_REG_INDEX(offset + read));
+		begin = STM32_BKP_REG_BYTE_INDEX(offset + read);
+		to_copy = MIN(STM32_BKP_REG_BYTES - begin, size - read);
+		bytecpy(data + read, (uint8_t *)&reg + begin, to_copy);
+	}
+
+	return 0;
+}
+
+static int bbram_stm32_write(const struct device *dev, size_t offset, size_t size,
+			     const uint8_t *data)
+{
+	const struct bbram_stm32_config *config = dev->config;
+	uint32_t reg, begin, to_copy;
+
+	if (size < 1 || offset + size > config->size) {
+		return -EFAULT;
+	}
+
+	for (size_t written = 0; written < size; written += to_copy) {
+		reg = STM32_BKP_REG(STM32_BKP_REG_INDEX(offset + written));
+		begin = STM32_BKP_REG_BYTE_INDEX(offset + written);
+		to_copy = MIN(STM32_BKP_REG_BYTES - begin, size - written);
+		bytecpy((uint8_t *)&reg + begin, data + written, to_copy);
+		STM32_BKP_REG(STM32_BKP_REG_INDEX(offset + written)) = reg;
+	}
+
+	return 0;
+}
+
+static int bbram_stm32_get_size(const struct device *dev, size_t *size)
+{
+	const struct bbram_stm32_config *config = dev->config;
+
+	*size = config->size;
+
+	return 0;
+}
+
+static const struct bbram_driver_api bbram_stm32_driver_api = {
+	.read = bbram_stm32_read,
+	.write = bbram_stm32_write,
+	.get_size = bbram_stm32_get_size,
+};
+
+static int bbram_stm32_init(const struct device *dev)
+{
+	const struct bbram_stm32_config *config = dev->config;
+
+	if (!device_is_ready(config->parent)) {
+		LOG_ERR("Device %s is not ready", config->parent->name);
+		return -ENODEV;
+	}
```

Comment:
I'd like to open the discussion on this part of your PR which seems the most controversial. This isn't a blocking comment.

At first sight, this dependency will save few bytes of ROM when both COUNTER and BBRAM are required by the application. On the other hand this will waste ROM in the case only BBRAM is required by the application, as COUNTER has to be complied even if not used.

For stm32 counter_ll_timer and pwm drivers, which share a similar constraint of using same IP for 2 different drivers, it was chose to make them fully independent. This is a slightly different case as in this case their usage is mutually exclusive on the same instance which is not true on current case.

Ideally we should be able to use multi function devices in that case. But I'm not sure there is any example yet available so I won't block.

Reply:
> For stm32 counter_ll_timer and pwm drivers, which share a similar constraint of using same IP for 2 different drivers, it was chose to make them fully independent. This is a slightly different case as in this case their usage is mutually exclusive on the same instance which is not true on current case.

We could add parent node for counter and backup_regs which would be responsible for basic initialization (clocks, interrupts), but it would be unnatural in my opinion.

It would be nice if we could implement both API (rtc and bbram) in one driver, with one init function, but I don't know if it's possible.